python/ray/data/_internal/planner/plan_udf_map_op.py

```diff
@@ -193,45 +210,94 @@ def _validate_batch_output(batch: Block) -> None:
 def _generate_transform_fn_for_map_batches(
     fn: UserDefinedFunction,
 ) -> MapTransformCallable[DataBatch, DataBatch]:
-    def transform_fn(
-        batches: Iterable[DataBatch], _: TaskContext
-    ) -> Iterable[DataBatch]:
-        for batch in batches:
-            try:
-                if (
-                    not isinstance(batch, collections.abc.Mapping)
-                    and BlockAccessor.for_block(batch).num_rows() == 0
-                ):
-                    # For empty input blocks, we directly ouptut them without
-                    # calling the UDF.
-                    # TODO(hchen): This workaround is because some all-to-all
-                    # operators output empty blocks with no schema.
-                    res = [batch]
-                else:
-                    res = fn(batch)
-                    if not isinstance(res, GeneratorType):
-                        res = [res]
-            except ValueError as e:
-                read_only_msgs = [
-                    "assignment destination is read-only",
-                    "buffer source array is read-only",
-                ]
-                err_msg = str(e)
-                if any(msg in err_msg for msg in read_only_msgs):
-                    raise ValueError(
-                        f"Batch mapper function {fn.__name__} tried to mutate a "
-                        "zero-copy read-only batch. To be able to mutate the "
-                        "batch, pass zero_copy_batch=False to map_batches(); "
-                        "this will create a writable copy of the batch before "
-                        "giving it to fn. To elide this copy, modify your mapper "
-                        "function so it doesn't try to mutate its input."
-                    ) from e
+    if inspect.iscoroutinefunction(fn):
+        # UDF is a callable class with async generator `__call__` method.
+        def transform_fn(
+            input_iterable: Iterable[DataBatch], _: TaskContext
+        ) -> Iterable[DataBatch]:
+            # Use a queue to store results from async generator calls.
+            # In the main event loop, we will put results into this queue
+            # from async generator, and yield them from the queue as they
+            # become available.
+            result_queue = queue.Queue()
+
+            async def process_batch(batch: DataBatch):
+                output_batch_iterator = await fn(batch)
+                # As soon as results become available from the async generator,
+                # put them into the result queue so they can be yielded.
+                async for output_row in output_batch_iterator:
+                    result_queue.put(output_row)
+
+            async def process_all_batches():
+                tasks = [asyncio.create_task(process_batch(x)) for x in input_iterable]
+                for task in asyncio.as_completed(tasks):
```

Comment:
`as_completed` doesn't preserve the order. Shouldn't use it when `preserve_order` is set.